File: R/radiationModel-internals.R
#' Create a monthly sequence of dates
#'
#' @param t_now Character, today date.
#' @param t_hor Character, horizon date.
#' @param last_day Logical. When `TRUE` the last day will be treated as conditional variance otherwise not.
#' @examples
#' t_now <- "2022-01-01"
#' t_hor <- "2022-03-24"
#' create_monthly_sequence("2022-01-01", "2022-03-24")
#' create_monthly_sequence("2022-01-01", "2022-03-24", TRUE)
#'
#' @keywords internal integral radiationModel
#' @export
#' @noRd
create_monthly_sequence <- function(t_now, t_hor, last_day = FALSE){
  # Convert in dates
  t_now <- as.Date(t_now)
  t_hor <- as.Date(t_hor)
  # Compute year and months for reference dates
  # Standard dates
  t_start <- as.Date(paste0(lubridate::year(t_now), "-", lubridate::month(t_now), "-01"))
  t_end <- as.Date(paste0(lubridate::year(t_hor), "-", lubridate::month(t_hor), "-01"))
  # Sequence of dates
  dates_hor <- dates_now <- seq.Date(t_start, t_end, by = "1 month")
  n_months <- length(dates_now)
  dates_now <- c(t_now, dates_now[-1] - lubridate::day(dates_now[-1]))
  dates_hor <- c(dates_hor[-n_months] - lubridate::day(dates_hor[-n_months]) + lubridate::days(lubridate::days_in_month(dates_hor[-n_months])), t_hor)
  n_of_day <- as.numeric(difftime(dates_hor, dates_now, units = "days"))
  df_dates <- data.frame(Year = lubridate::year(dates_now), Month = lubridate::month(dates_hor), n = n_of_day)
  df_dates$N <- df_dates$n + number_of_day(t_now) + c(0, cumsum(lag(df_dates$n, 1)[-1]))
  df_dates$n <- df_dates$N - df_dates$n
  df_dates$tau <- as.numeric(difftime(t_hor, t_now, units = "days")) + df_dates$n[1]
  if (last_day) {
    df_last <- dplyr::mutate(tail(df_dates, 1), n = tau - 1)
    df_dates[nrow(df_dates),]$N <- df_dates[nrow(df_dates),]$N - 1
    df_dates <- dplyr::bind_rows(df_dates, df_last)
  }
  attr(df_dates, "last_day") <- last_day
  return(df_dates)
}

# Martingale estimation for mean-reverson
martingale_method_seasonal <- function(Yt, Yt_bar, e_mu = 0){
  # Dataset
  data <- dplyr::tibble(Yt = Yt, Yt_bar = Yt_bar, e_mu = e_mu)
  data$n <- 1:nrow(data)
  # Quadratic variation
  data$dYt2 <- (lag(data$Yt,1) - lag(data$Yt, 2))^2
  # Seeasonal variance from quadratic variation
  seasonal_variance <- seasonalModel$new()
  seasonal_variance$fit("dYt2 ~ 1", data = data)
  # Estimated seasonal variance
  data$sigma2_bar <- seasonal_variance$predict(1:nrow(data)-1)
  # Martingale estimation
  data$Y_est_L1 <- (lag(data$Yt_bar, 1) - lag(data$Yt, 1)) / data$sigma2_bar
  # Differences from seasonal mean
  data$dYt <-  data$Yt - data$Yt_bar
  data$dYt_L1 <- lag(data$Yt, 1) - lag(data$Yt_bar, 1) - lag(data$e_mu, 1)
  data <- na.omit(data)
  a_n <- sum(data$Y_est_L1 * data$dYt) / sum(data$Y_est_L1 * data$dYt_L1)
  -log(a_n)
}
# Reparametrization for seasonal variance
reparam_seasonal_function <- function(par, theta, omega){
  # Original parameters
  a0 <- par[1]
  a1 <- par[2]
  a2 <- par[3]
  # Correction for long term variance
  c0_long <- a0 * 2 * theta
  c1_long <- a1 * 2 * theta - omega * a2
  c2_long <- a2 * 2 * theta + omega * a1
  # Integral parameters
  gamma0_long <-  c0_long / (2 * theta)
  gamma1_long <- (c1_long * 2 * theta + c2_long * omega) / (4 * theta^2 + omega^2)
  gamma2_long <- (c2_long * 2 * theta - c1_long * omega) / (4 * theta^2 + omega^2)

  # Correction for short term variance
  alpha  <- 1 - exp(-2 * theta) * cos(omega)
  beta  <- exp(-2 * theta) * sin(omega)
  detM  <- alpha^2 + beta^2
  c0 <- (2 * theta * a0) / (1 - exp(-2 * theta))
  c1 <- ((2 * theta * alpha + omega * beta) * a1 + (2 * theta * beta - omega * alpha) * a2) / detM
  c2 <- ((omega * alpha - 2 * theta * beta) * a1 + (omega * beta + 2 * theta * alpha) * a2) / detM
  # Integral parameters
  gamma0 <-  c0 / (2 * theta)
  gamma1 <- (c1 * 2 * theta + c2 * omega) / (4 * theta^2 + omega^2)
  gamma2 <- (c2 * 2 * theta - c1 * omega) / (4 * theta^2 + omega^2)
  structure(
    list(
      alpha = alpha,
      beta = beta,
      detM = detM,
      a_ = c(a0 = a0, a1 = a1, a2 = a2),
      c_ = c(c0 = c0, c1 = c1, c2 = c2),
      c_long = c(c0 = c0_long, c1 = c1_long, c2 = c2_long),
      gamma = c(gamma0 = gamma0, gamma1 = gamma1, gamma2 = gamma2),
      gamma_long = c(gamma0 = gamma0_long, gamma1 = gamma1_long, gamma2 = gamma2_long)
    )
  )
}
# Numeric integral for the square root of the sigma_s * exp(-theta)
integral_sigma_numeric <- function(theta, par, omega = 2*base::pi/365){
  seasonal_function <- function(t) par[1] + par[2] * sin(omega * t) + par[3] * cos(omega * t)
  integrand <- function(tau, T_) sqrt(seasonal_function(tau) * exp(-2*theta*(T_-tau)))
  function(t, s, T_){
    result <- c()
    for(i in 1:length(T_)){
      result[i] <- integrate(integrand, lower = t[i], upper = s[i], T_ = T_[i])$value
    }
    return(result)
  }
}
# Formula integral for sigma_s^2 * exp(-2theta)
integral_sigma2_formula <- function(theta, par, omega = 2*base::pi/365){
  # Functions
  f0 <- function(t, s, T_) exp(-2 * theta * (T_ - s)) - exp(-2 * theta * (T_ - t))
  f1 <- function(t, s, T_) exp(-2 * theta * (T_ - s)) * sin(omega * s) - exp(-2 * theta * (T_ - t)) * sin(omega * t)
  f2 <- function(t, s, T_) exp(-2 * theta * (T_ - s)) * cos(omega * s) - exp(-2 * theta * (T_ - t)) * cos(omega * t)

  function(t, s, T_){
    result <- par[1] * f0(t, s, T_) + par[2] * f1(t, s, T_) + par[3] * f2(t, s, T_)
    return(result)
  }
}

File: solarrpy/radiationModel_internals.py
import warnings
import numpy as np
import pandas as pd
from scipy.integrate import quad

from .zzz import number_of_day
from .seasonalModel import SeasonalModel

"""
Solar model helper functions — Python translation.

Functions:
    create_monthly_sequence       — monthly date grid between t_now and t_hor
    martingale_method_seasonal    — mean-reversion θ via martingale estimator
    reparam_seasonal_function     — reparametrise seasonal variance to continuous time
    integral_sigma_numeric        — numerical integral of √(σ̄(s) · e^{-θ(T-s)})
    integral_sigma2_formula       — closed-form integral of σ̄²(s) · e^{-2θ(T-s)}

Version: 1.0.0
"""

def create_monthly_sequence(t_now, t_hor, last_day=False):
    t_now = pd.Timestamp(t_now)
    t_hor = pd.Timestamp(t_hor)

    # First day of the month for t_now and t_hor
    t_start = t_now.replace(day=1)
    t_end   = t_hor.replace(day=1)

    # Monthly sequence of first-of-month dates — mirrors seq.Date(..., by="1 month")
    months = pd.date_range(t_start, t_end, freq="MS")
    
    # dates_now: t_now, then last day of previous month for each subsequent first-of-month
    dates_now = [t_now] + [m - pd.Timedelta(days=m.day) for m in months[1:]]
    
    # dates_hor: last day of each month for all-but-last, then t_hor
    dates_hor = [m + pd.offsets.MonthEnd(0) for m in months[:-1]] + [t_hor]

    # Number of days in each sub-interval — mirrors difftime(dates_hor, dates_now)
    n_of_day = np.array([(h - s).days for s, h in zip(dates_now, dates_hor)], dtype=float)

    # Build base dataframe — Year from dates_now, Month from dates_hor
    df = pd.DataFrame({
        "Year":  [d.year  for d in dates_now],
        "Month": [d.month for d in dates_hor],
        "n":     n_of_day,
    })
    
    # Mirror R's c(0, cumsum(lag(df_dates$n, 1)[-1])): cumulative end-of-sub-interval
    # offsets computed from the LAGGED lengths (previous values), not the leading slice.
    # The original n_of_day[1:] form was off-by-one and pushed sub-interval endpoints
    # beyond tau, causing exp(-2θ(T−s)) in integral_sigma2_formula to blow up for
    # any horizon > 1 day. See R radiationModel-internals.R:30.
    df["N"] = df["n"] + number_of_day(t_now) + np.concatenate([[0], np.cumsum(n_of_day[:-1])])

    # n = N - n  (start index of each sub-interval)
    df["n"] = df["N"] - df["n"]

    # tau = total days t_now→t_hor + n[0]  (constant column)
    df["tau"] = (t_hor - t_now).days + df["n"].iloc[0]

    # Optional last-day split
    if last_day:
        last_row = df.iloc[[-1]].copy()
        last_row["n"] = last_row["tau"] - 1          # mirrors mutate(n = tau - 1)
        df.iloc[-1, df.columns.get_loc("N")] -= 1    # trim last row's N by 1
        df = pd.concat([df, last_row], ignore_index=True)

    df.attrs["last_day"] = last_day
    return df[['Year', 'Month', 'n', 'N', 'tau']]

# ---------------------------------------------------------------------------
# martingale_method_seasonal
# ---------------------------------------------------------------------------
def martingale_method_seasonal(Yt, Yt_bar, e_mu=0):
    Yt = np.asarray(Yt)
    Yt_bar = np.asarray(Yt_bar)

    if np.isscalar(e_mu):
        e_mu = np.full_like(Yt, e_mu)
    else:
        e_mu = np.asarray(e_mu)

    # ── CHANGED: use global sequential counter 1..N, matching R's
    #    data$n <- 1:nrow(data)
    #    The Fourier OLS for sigma2_bar must see n growing monotonically
    #    across the full training window, not resetting each January.
    n_global = np.arange(1, len(Yt) + 1, dtype=float)

    # Quadratic variation
    dYt2 = (Yt - pd.Series(Yt).shift(1).values) ** 2

    data_df = pd.DataFrame({'dYt2': dYt2, 'n': n_global})
    fit_df = data_df.dropna(subset=['dYt2']).copy()

    seasonal_variance = SeasonalModel()
    seasonal_variance.fit(
        data=fit_df,
        target_col='dYt2',
        time_col='n',
        include_intercept=True
    )

    # Predict sigma2_bar at global n − 1 to match R's predict(1:nrow(data) - 1)
    pred_df = pd.DataFrame({'n': n_global - 1})
    sigma2_bar = seasonal_variance.predict(data=pred_df, time_col='n')

    # Martingale components — unchanged
    Y_est_L1 = (pd.Series(Yt_bar).shift(1).values - pd.Series(Yt).shift(1).values) / sigma2_bar.values
    dYt = Yt - Yt_bar
    dYt_L1 = (pd.Series(Yt).shift(1).values
               - pd.Series(Yt_bar).shift(1).values
               - pd.Series(e_mu).shift(1).values)

    df = pd.DataFrame({
        "Y_est_L1": Y_est_L1,
        "dYt":      dYt,
        "dYt_L1":   dYt_L1,
    }).dropna()

    a_n = np.sum(df["Y_est_L1"] * df["dYt"]) / np.sum(df["Y_est_L1"] * df["dYt_L1"])
    return -np.log(a_n)

# ---------------------------------------------------------------------------
# reparam_seasonal_function
# ---------------------------------------------------------------------------
def reparam_seasonal_function(par, theta, omega=2 * np.pi / 365):
    """
    Reparametrization for seasonal variance. Maps OLS parameters to continuous-time parameters.
    """
    b0, b1, b2 = par[0], par[1], par[2]
    
    # Correction for long term variance
    c0_long = b0 * 2 * theta
    c1_long = b1 * 2 * theta - omega * b2
    c2_long = b2 * 2 * theta + omega * b1
    
    gamma0_long = c0_long / (2 * theta)
    gamma1_long = (c1_long * 2 * theta + c2_long * omega) / (4 * theta**2 + omega**2)
    gamma2_long = (c2_long * 2 * theta - c1_long * omega) / (4 * theta**2 + omega**2)

    # Correction for short term variance
    phi1 = 1 - np.exp(-2 * theta) * np.cos(omega)
    phi2 = np.exp(-2 * theta) * np.sin(omega)
    detM = phi1**2 + phi2**2
    
    c0 = (2 * theta * b0) / (1 - np.exp(-2 * theta))
    c1 = ((2 * theta * phi1 + omega * phi2) * b1 + (2 * theta * phi2 - omega * phi1) * b2) / detM
    c2 = ((omega * phi1 - 2 * theta * phi2) * b1 + (omega * phi2 + 2 * theta * phi1) * b2) / detM
    
    gamma0 = c0 / (2 * theta)
    gamma1 = (c1 * 2 * theta + c2 * omega) / (4 * theta**2 + omega**2)
    gamma2 = (c2 * 2 * theta - c1 * omega) / (4 * theta**2 + omega**2)
    
    return {
        "phi1": phi1,
        "phi2": phi2,
        "detM": detM,
        "b_": np.array([b0, b1, b2]),
        "c_": np.array([c0, c1, c2]),
        "c_long": np.array([c0_long, c1_long, c2_long]),
        "gamma": np.array([gamma0, gamma1, gamma2]),
        "gamma_long": np.array([gamma0_long, gamma1_long, gamma2_long])
    }

# ---------------------------------------------------------------------------
# integral_sigma_numeric
# ---------------------------------------------------------------------------
def integral_sigma_numeric(theta, par, omega=2 * np.pi / 365):
    """
    Returns a function that computes the numerical integral for the square root of sigma_s * exp(-theta).
    """
    c0, c1, c2 = par[0], par[1], par[2]
    
    def seasonal_function(tau):
        return c0 + c1 * np.sin(omega * tau) + c2 * np.cos(omega * tau)
        
    def integrand(tau, T_val):
        val = seasonal_function(tau) * np.exp(- 2 * theta * (T_val - tau))
        # Protect against negative values inside the sqrt due to numerical drift
        return np.sqrt(np.maximum(val, 0))
        
    def integrator(t, s, T_):
        t = np.atleast_1d(t)
        s = np.atleast_1d(s)
        T_ = np.atleast_1d(T_)
        
        result = np.zeros(len(T_))
        for i in range(len(T_)):
            res, _ = quad(integrand, t[i], s[i], args=(T_[i],))
            result[i] = res
        return result
        
    return integrator

# ---------------------------------------------------------------------------
# integral_sigma2_formula
# ---------------------------------------------------------------------------
def integral_sigma2_formula(theta, par, omega=2 * np.pi / 365):
    """
    Returns a vectorized function computing the exact formula integral for sigma_s^2 * exp(-2theta).
    """
    g0, g1, g2 = par[0], par[1], par[2]
    
    def f0(t, s, T_):
        return np.exp(-2 * theta * (T_ - s)) - np.exp(-2 * theta * (T_ - t))
        
    def f1(t, s, T_):
        return np.exp(-2 * theta * (T_ - s)) * np.sin(omega * s) - np.exp(-2 * theta * (T_ - t)) * np.sin(omega * t)
        
    def f2(t, s, T_):
        return np.exp(-2 * theta * (T_ - s)) * np.cos(omega * s) - np.exp(-2 * theta * (T_ - t)) * np.cos(omega * t)

    def integrator(t, s, T_):
        return g0 * f0(t, s, T_) + g1 * f1(t, s, T_) + g2 * f2(t, s, T_)
        
    return integrator